specs/protocol/exec-engine.md

```diff
@@ -156,6 +156,98 @@ The two base fee values and their respective scalars can be accessed in two inte
+| `intercept`       | `int32`   | Intercept constant, scaled by `1e6` (can be negative)             |
+| `fastlzCoef`      | `int32`   | FastLZ coefficient, scaled by `1e6` (can be negative)             |
+| `txSizeCoef`      | `int32`   | Transaction size coefficient, scaled by `1e6` (can be negative)   |
+
+Where:
+
+- the `l1CostSigned` calculation is an unlimited precision signed integer computation, with the result in Wei and
+  having `int256` range.
+
+- the final `l1Cost` computation is an unlimited precision unsigned integer computation, with the result in Wei and
+  having `uint256` range.
+
+##### L1-Cost linear regression details
+
+The `intercept`, `fastlzCoef`, and `txSizeCoef` constants are calculated by linear regression using a dataset
+of previous L2 transactions. The dataset is generated by iterating over all transactions in a given time range, and
+performing the following actions. For each transaction:
+
+1. Calculate the RLP-encoded transaction payload. Record the length of this payload as `txSize`.
+2. Compress the payload using FastLZ. Record the length of the compressed payload as `fastlzSize`.
+3. Compress the payload using a "best estimate" compression, defined as:
+
+- Create two zlib streams, both using the maximum compression level.
+- Bootstrap one of the streams with at least 64kb of transaction data (after compression).
+- Write each transaction to both streams (flushing after each write), and record the size increase
+    of the larger buffer as `bestEstimateSize`.
+- Each time the larger buffer grows larger than 128kb, empty the stream and swap the buffers.
+
+Once this dataset is generated, a linear regression can be calculated using the `bestEstimateSize` as
+the dependent variable and `fastlzSize` and `txSize` as the independent variables.
+
+The following Python code can be used to calculate the linear regression:
+
+```python
+import numpy as np
+from sklearn.linear_model import LinearRegression
+
+# Dataset format is:
+#   data               = [record...] (one or more records, concatenated)
+#   record             = bestEstimateSize ++ fastlzSize ++ txSize
+#   bestEstimateSize   = little-endian uint32
+#   fastlzSize         = little-endian uint32
+#   txSize             = little-endian uint32
+
+dt = np.dtype([('best', '<u4'), ('fastlz', '<u4'), ('length', '<u4')])
+data = np.fromfile('./data.bin', dtype=dt)
+input_array = np.array(data.tolist())
+
+x = np.delete(input_array, [0], 1)
+y = input_array[:, 0]
+model = LinearRegression().fit(x, y)
+print(f'model: {model.intercept_} {model.coef_}')
+```
+
+As an example, we generated a dataset from all transactions on Optimism Mainnet in October 2023,
+and the resulting linear regression was:
+
+`-27.321890037208703 + 1.03146206*fastlzSize - 0.08866427*txSize`
+
+Scaling these values by `1e6` gives the constants used in the L1-Cost fee estimator:
+
+- `intercept = -27_321_890`
```

Comment:
How much precision do we want / need here?

Reply:
we are currently proposing signed 32-bit integers, as we don't see a need to go above 2^31 for any of these values